src/libsyntax/parse/parser.rs

```diff
@@ -3154,13 +3154,47 @@ impl<'a> Parser<'a> {
         // Parse: `for <src_pat> in <src_expr> <src_loop_block>`
 
         let pat = self.parse_pat()?;
-        self.expect_keyword(keywords::In)?;
```

Comment:
This can be done in much simpler way without parser snapshots (please don't overuse the trick from https://github.com/rust-lang/rust/pull/42578).

```rust
if !self.eat_keyword(keywords::In) {
    span_err(/*missing `in`*/) // Non-fatal error
}
let expr = self.parse_expr_res(...); // Then parse as usual as if `in` were written
```